server/src/arcs-master-app.ts

```diff
@@ -32,41 +40,128 @@ class ArcsMasterApp extends AppBase {
-      const container: Container | null = await cloud.containers().find(fingerprint);
-      if (container != null) {
-        const url = container.url();
-        if (url === 'pending') {
-          res.send('Creation still in progress, refresh for status.');
+      const disk = await cloud.disks().find(fingerprint);
+      if (disk) {
+        const attached = await disk.isAttached();
+        if (attached) {
+          return;
+        }
+        const container: Container | null = await cloud.containers().find(fingerprint);
+        if (container != null) {
+          await disk.mount(req.params.rewrappedKey, await container.node());
+          await fetch("http://" + container.url() + "/unlock");
+          return;
         } else {
-          res.send(url);
+          // TODO: if container not found, deploy a new one?
         }
-      } else {
-        res.send('Container ' + fingerprint + ' not found.');
       }
+    } catch (e) {
+      console.log(e);
+    }
+  }
 
+  async findDeployment(req, res, next) {
+    console.log("fingerprint is " + req.params.fingerprint);
+    const fingerprint = this.gcpSafeIdentifier(req.params.fingerprint);
+    console.log("gcp safe fingerprint " + fingerprint);
+    try {
+      const cloud = CloudManager.forGCP();
+      const disk = await cloud.disks().find(fingerprint);
+      if (disk) {
+        const attached = await disk.isAttached();
+        const wrappedKey  = await disk.wrappedKeyFor(fingerprint);
+
+        if (attached) {
+          const container: Container | null = await cloud.containers().find(fingerprint);
+          if (container != null) {
+            const status = container.status();
+            if (status !== 'Running') {
+              res.send('{"id": "' + fingerprint + '", "status": "pending"}');
+            } else {
+              res.send('{"id": "' + fingerprint + '", "status": "attached", "url" : "' + container.url() + '"}');
+            }
+            return;
+          }
+          res.send('{"id": "' + fingerprint + '", "status": "container not found"}');
+          return;
+        } else {
+          res.send('{"id": "' + fingerprint + '", "status": "detached", "key": "' + wrappedKey + '"}');
+          return;
+        }
+      } else {
+        res.send('{"id": "' + fingerprint + '", "status": "disk not found"}');
+        return;
+      }
     } catch (e) {
       console.log(e);
     }
   }
 
   async deploy(req, res, next) {
     const fingerprint = this.gcpSafeIdentifier(req.params.fingerprint);
-    const wrappedkey = req.params.wrappedkey;
+    const wrappedKey = req.params.wrappedKey;
+    const rewrappedKey = req.params.rewrappedKey;
+
     try {
       const cloud = CloudManager.forGCP();
-      const disk = await cloud.disks().create(fingerprint, wrappedkey);
+      const disk = await cloud.disks().create(fingerprint, wrappedKey, rewrappedKey);
       console.log("Disk successfully created with id " + disk.id());
```

Comment:
Probably want to null-check (or however errors are reported from `create` and `deploy`) here and line 143 to make sure disk and container were created. Or, do they throw an exception which would hit line 146?

Similar for `mount` implementation below.

Reply:
Currently, it's exception based. I added an intermediate check and a TODO to add better error handling.